File: aman-captain/lib/features/trip/domain/trip.dart
import 'package:equatable/equatable.dart';

enum TripStatus {
  requested, searching, accepted, arrived, inProgress, completed, cancelled, noCaptainFound,
}

TripStatus tripStatusFromString(String value) {
  switch (value) {
    case 'requested': return TripStatus.requested;
    case 'searching': return TripStatus.searching;
    case 'accepted': return TripStatus.accepted;
    case 'arrived': return TripStatus.arrived;
    case 'in_progress': return TripStatus.inProgress;
    case 'completed': return TripStatus.completed;
    case 'cancelled': return TripStatus.cancelled;
    case 'no_captain_found': return TripStatus.noCaptainFound;
    default: return TripStatus.requested;
  }
}

class TripCustomer extends Equatable {
  const TripCustomer({required this.id, this.fullName, required this.phone});

  final int id;
  final String? fullName;
  final String phone;

  factory TripCustomer.fromJson(Map<String, dynamic> json) => TripCustomer(
        id: json['id'] as int,
        fullName: json['full_name'] as String?,
        phone: json['phone'] as String,
      );

  @override
  List<Object?> get props => [id, fullName, phone];
}

class TripLocation extends Equatable {
  const TripLocation({required this.address, required this.lat, required this.lng});

  final String address;
  final double lat;
  final double lng;

  factory TripLocation.fromJson(Map<String, dynamic> json) => TripLocation(
        address: json['address'] as String,
        lat: (json['lat'] as num).toDouble(),
        lng: (json['lng'] as num).toDouble(),
      );

  @override
  List<Object?> get props => [address, lat, lng];
}

class Trip extends Equatable {
  const Trip({
    required this.id,
    required this.tripCode,
    required this.status,
    required this.serviceName,
    required this.pickup,
    this.dropoff,
    this.customer,
    this.estimatedPrice,
    this.finalPrice,
    required this.currency,
    this.distanceKm,
    this.durationMin,
  });

  final int id;
  final String tripCode;
  final TripStatus status;
  final String serviceName;
  final TripLocation pickup;
  final TripLocation? dropoff;
  final TripCustomer? customer;
  final double? estimatedPrice;
  final double? finalPrice;
  final String currency;
  final double? distanceKm;
  final int? durationMin;

  factory Trip.fromJson(Map<String, dynamic> json) {
    return Trip(
      id: json['id'] as int,
      tripCode: json['trip_code'] as String,
      status: tripStatusFromString(json['status'] as String),
      serviceName: (json['service']?['name'] as String?) ?? '',
      pickup: TripLocation.fromJson(json['pickup'] as Map<String, dynamic>),
      dropoff: json['dropoff'] != null ? TripLocation.fromJson(json['dropoff'] as Map<String, dynamic>) : null,
      customer: json['customer'] != null && (json['customer'] as Map).isNotEmpty
          ? TripCustomer.fromJson(json['customer'] as Map<String, dynamic>)
          : null,
      estimatedPrice: (json['estimated_price'] as num?)?.toDouble(),
      finalPrice: (json['final_price'] as num?)?.toDouble(),
      currency: json['currency'] as String? ?? 'MRU',
      distanceKm: (json['distance_km'] as num?)?.toDouble(),
      durationMin: json['duration_min'] as int?,
    );
  }

  bool get isActive => status == TripStatus.accepted || status == TripStatus.arrived || status == TripStatus.inProgress;

  @override
  List<Object?> get props => [id, status, tripCode];
}

/// عرض رحلة جديد يصل عبر البث اللحظي (Reverb) — راجع NewTripOffer في aman-backend
class TripOffer extends Equatable {
  const TripOffer({
    required this.tripId,
    required this.pickupAddress,
    required this.pickupLat,
    required this.pickupLng,
    required this.estimatedPrice,
    required this.currency,
    required this.offerTimeoutSeconds,
  });

  final int tripId;
  final String pickupAddress;
  final double pickupLat;
  final double pickupLng;
  final double estimatedPrice;
  final String currency;
  final int offerTimeoutSeconds;

  factory TripOffer.fromJson(Map<String, dynamic> json) => TripOffer(
        tripId: json['trip_id'] as int,
        pickupAddress: json['pickup_address'] as String,
        pickupLat: (json['pickup_lat'] as num).toDouble(),
        pickupLng: (json['pickup_lng'] as num).toDouble(),
        estimatedPrice: (json['estimated_price'] as num).toDouble(),
        currency: json['currency'] as String,
        offerTimeoutSeconds: json['offer_timeout_seconds'] as int? ?? 15,
      );

  @override
  List<Object?> get props => [tripId];
}

File: aman-backend/app/Models/Trip.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\HasOne;

class Trip extends Model
{
    use HasFactory;

    // ==== حالات الرحلة ====
    public const STATUS_REQUESTED = 'requested';
    public const STATUS_SEARCHING = 'searching';
    public const STATUS_ACCEPTED = 'accepted';
    public const STATUS_ARRIVED = 'arrived';
    public const STATUS_IN_PROGRESS = 'in_progress';
    public const STATUS_COMPLETED = 'completed';
    public const STATUS_CANCELLED = 'cancelled';
    public const STATUS_NO_CAPTAIN_FOUND = 'no_captain_found';

    // ==== أنماط الرحلة ====
    public const MODE_INSTANT = 'instant';
    public const MODE_SCHEDULED = 'scheduled';
    public const MODE_OPEN = 'open';

    protected $fillable = [
        'uuid', 'trip_code', 'customer_id', 'captain_id', 'vehicle_id',
        'service_id', 'city_id', 'trip_mode', 'status',
        'pickup_address', 'pickup_lat', 'pickup_lng',
        'dropoff_address', 'dropoff_lat', 'dropoff_lng',
        'scheduled_at', 'distance_km', 'duration_min',
        'estimated_price', 'final_price', 'currency',
        'pricing_rule_id', 'commission_rule_id', 'commission_amount',
        'cancelled_by_type', 'cancelled_by_id', 'cancel_reason',
        'requested_at', 'accepted_at', 'arrived_at', 'started_at', 'completed_at', 'cancelled_at',
    ];

    protected $casts = [
        'scheduled_at' => 'datetime',
        'requested_at' => 'datetime',
        'accepted_at' => 'datetime',
        'arrived_at' => 'datetime',
        'started_at' => 'datetime',
        'completed_at' => 'datetime',
        'cancelled_at' => 'datetime',
        'estimated_price' => 'decimal:2',
        'final_price' => 'decimal:2',
        'commission_amount' => 'decimal:2',
    ];

    protected static function booted(): void
    {
        static::creating(function (Trip $trip) {
            $trip->uuid ??= (string) \Illuminate\Support\Str::uuid();
            $trip->trip_code ??= 'TR-' . strtoupper(\Illuminate\Support\Str::random(8));
            $trip->requested_at ??= now();
        });
    }

    // ==== العلاقات ====
    public function customer(): BelongsTo
    {
        return $this->belongsTo(Customer::class);
    }

    public function captain(): BelongsTo
    {
        return $this->belongsTo(Captain::class);
    }

    public function vehicle(): BelongsTo
    {
        return $this->belongsTo(Vehicle::class);
    }

    public function service(): BelongsTo
    {
        return $this->belongsTo(Service::class);
    }

    public function city(): BelongsTo
    {
        return $this->belongsTo(City::class);
    }

    public function pricingRule(): BelongsTo
    {
        return $this->belongsTo(PricingRule::class);
    }

    public function deliveryDetails(): HasOne
    {
        return $this->hasOne(TripDeliveryDetail::class);
    }

    public function airportDetails(): HasOne
    {
        return $this->hasOne(TripAirportDetail::class);
    }

    public function statusHistory(): HasMany
    {
        return $this->hasMany(TripStatusHistory::class)->orderBy('created_at');
    }

    public function dispatchAttempts(): HasMany
    {
        return $this->hasMany(TripDispatchAttempt::class);
    }

    public function locationPings(): HasMany
    {
        return $this->hasMany(TripLocationPing::class);
    }

    public function rating(): HasOne
    {
        return $this->hasOne(Rating::class);
    }

    public function walletTransactions(): HasMany
    {
        return $this->hasMany(WalletTransaction::class);
    }

    // ==== قواعد عمل مساعدة ====
    public function isCancellable(): bool
    {
        return in_array($this->status, [
            self::STATUS_REQUESTED, self::STATUS_SEARCHING,
            self::STATUS_ACCEPTED, self::STATUS_ARRIVED,
        ], true);
    }

    public function isActive(): bool
    {
        return in_array($this->status, [
            self::STATUS_ACCEPTED, self::STATUS_ARRIVED, self::STATUS_IN_PROGRESS,
        ], true);
    }
}
